prov/util/src/util_attr.c

```diff
@@ -721,11 +722,52 @@ int ofi_check_tx_attr(const struct fi_provider *prov,
 	return 0;
 }
 
-int ofi_check_info(const struct util_prov *util_prov, uint32_t api_version,
-		   const struct fi_info *user_info)
+/* if there are multiple fi_info in the provider */
+int ofi_check_info_ex(const struct util_prov *util_prov, uint32_t api_version,
+		      const struct fi_info *user_info,
+		      struct fi_info **ret_info)
 {
 	const struct fi_info *prov_info = util_prov->info;
 	const struct fi_provider *prov = util_prov->prov;
+	struct fi_info *info = NULL;
+	int ret, first = 1, res = -FI_ENODATA;
+
+	for ( ; prov_info; prov_info = prov_info->next) {
+		ret = ofi_check_info(util_prov, prov_info,
+				     api_version, user_info);
+	    	if (ret)
+			continue;
+		if (ret_info) {
+			info = fi_dupinfo(prov_info);
+			if (first) {
+				if (!info)
+					goto first;
+				*ret_info = info;
+				first = 0;
+			}
+			if (!info)
+				goto other;
+			info = info->next;
+		}
+		/* found one valid fi_info at least */
+		res = FI_SUCCESS;
+	}
+
+	return res;
+other:
+	fi_freeinfo(*ret_info);
+first:
+	FI_INFO(prov, FI_LOG_CORE,
+		"cannot copy info\n");
+	return -FI_ENOMEM;
+}
```

Comment:
ofi_check_info is a validation checking function.  ofi_check_info_ex has the side effect of also creating a list of fi_info structures that must then be altered by another call.  Unless ofi_check_info_ex is called with a NULL user_info parameter, in which case the dupinfo is skipped.  This is somewhat confusing, plus the returned info list was never modified (see comment below).  Please see if this approach would be cleaner:

1. extend ofi_check_info() as proposed - it verifies a single user_info vs a prov_info.
2. add ofi_prov_check_info() - verifies a single user_info against a provider info list.  This is ofi_check_info_ex() with the ret_info parameter removed.  (I'm not sold on the name).
3. Update util_getinfo() to handle a provider info list.  Update the verbs provider to use that call.